### PR TITLE
initial commit testcontainers-rabbitmq

### PR DESCRIPTION
**Pull request by dmfrey**

### CODE REVIEW

**Comment by pivotal-issuemaster** [comment]
@dmfrey Please sign the [Contributor License Agreement](https://cla.pivotal.io/sign/spring?repositoryId=spring-projects/spring-integration-samples&pullRequestId=280)!

[Click here](https://cla.pivotal.io/sync/spring?repositoryId=spring-projects/spring-integration-samples&pullRequestId=280) to manually synchronize the status of this Pull Request.

See the [FAQ](https://cla.pivotal.io/about) for frequently asked questions.

**Comment by pivotal-issuemaster** [comment]
@dmfrey Thank you for signing the [Contributor License Agreement](https://cla.pivotal.io/sign/spring?repositoryId=spring-projects/spring-integration-samples&pullRequestId=280)!

File: intermediate/testcontainers-rabbitmq/src/test/java/org/springframework/integration/samples/testcontainersrabbitmq/TestcontainersRabbitmqApplicationTests.java [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+import org.junit.jupiter.api.Test;
+import org.springframework.boot.test.context.SpringBootTest;
+import org.springframework.context.annotation.Import;
+
+@SpringBootTest
+@Import({ Receiver.class, IntegrationConfigTest.Config.class })
+class TestcontainersRabbitmqApplicationTests {
+
+	@Test
+	void contextLoads() {
```
Comment (artembilan) [inline intermediate/testcontainers-rabbitmq/src/test/java/org/springframework/integration/samples/testcontainersrabbitmq/TestcontainersRabbitmqApplicationTests.java]: Why do we need this one if it is void?

File: intermediate/testcontainers-rabbitmq/pom.xml [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+		</dependency>
+		<dependency>
+			<groupId>org.springframework.amqp</groupId>
+			<artifactId>spring-rabbit-test</artifactId>
+			<scope>test</scope>
+		</dependency>
+
+		<dependency>
+			<groupId>org.springframework.cloud</groupId>
+			<artifactId>spring-cloud-starter</artifactId>
```
[the remark this answers is not in the document]
Comment (dmfrey) [reply intermediate/testcontainers-rabbitmq/pom.xml]: It is required for the library that starts the embedded rabbitmq server

```

File: intermediate/testcontainers-rabbitmq/pom.xml [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+		</dependency>
+		<dependency>
+			<groupId>org.springframework.boot</groupId>
+			<artifactId>spring-boot-starter-json</artifactId>
+		</dependency>
+
+		<!-- Third-party dependencies -->
+		<dependency>
+			<groupId>org.projectlombok</groupId>
+			<artifactId>lombok</artifactId>
```
[the remark this answers is not in the document]
Comment (dmfrey) [reply intermediate/testcontainers-rabbitmq/pom.xml]: removed

File: intermediate/testcontainers-rabbitmq/pom.xml [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+	xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 https://maven.apache.org/xsd/maven-4.0.0.xsd">
+	<modelVersion>4.0.0</modelVersion>
+	<parent>
+		<groupId>org.springframework.boot</groupId>
+		<artifactId>spring-boot-starter-parent</artifactId>
+		<version>2.2.5.RELEASE</version>
+		<relativePath/> <!-- lookup parent from repository -->
+	</parent>
+	<groupId>org.springframework.integration.samples</groupId>
```
[the remark this answers is not in the document]
Comment (dmfrey) [reply intermediate/testcontainers-rabbitmq/pom.xml]: converted

File: intermediate/testcontainers-rabbitmq/src/main/java/org/springframework/integration/samples/testcontainersrabbitmq/IntegrationConfig.java [hunk 5]
```diff
@@ -0,0 +1,30 @@
+package org.springframework.integration.samples.testcontainersrabbitmq;
```
[the remark this answers is not in the document]
Comment (dmfrey) [reply intermediate/testcontainers-rabbitmq/src/main/java/org/springframework/integration/samples/testcontainersrabbitmq/IntegrationConfig.java]: added

File: intermediate/testcontainers-rabbitmq/src/main/java/org/springframework/integration/samples/testcontainersrabbitmq/IntegrationConfig.java [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+import org.springframework.context.annotation.Configuration;
+import org.springframework.integration.amqp.dsl.Amqp;
+import org.springframework.integration.annotation.IntegrationComponentScan;
+import org.springframework.integration.config.EnableIntegration;
+import org.springframework.integration.dsl.IntegrationFlow;
+
+@Slf4j
+@Configuration
+@EnableIntegration
+@IntegrationComponentScan
```
[the remark this answers is not in the document]
Comment (dmfrey) [reply intermediate/testcontainers-rabbitmq/src/main/java/org/springframework/integration/samples/testcontainersrabbitmq/IntegrationConfig.java]: removed

File: intermediate/testcontainers-rabbitmq/src/test/java/org/springframework/integration/samples/testcontainersrabbitmq/IntegrationConfigTest.java [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+import org.springframework.messaging.support.MessageBuilder;
+import org.springframework.test.context.junit.jupiter.SpringExtension;
+
+import java.util.Optional;
+import java.util.UUID;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.springframework.boot.test.context.SpringBootTest.WebEnvironment.NONE;
+
+@ExtendWith( SpringExtension.class )
```
[the remark this answers is not in the document]
Comment (dmfrey) [reply intermediate/testcontainers-rabbitmq/src/test/java/org/springframework/integration/samples/testcontainersrabbitmq/IntegrationConfigTest.java]: removed

File: intermediate/testcontainers-rabbitmq/src/test/java/org/springframework/integration/samples/testcontainersrabbitmq/IntegrationConfigTest.java [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+@SpringBootTest( webEnvironment = NONE )
```
[the remark this answers is not in the document]
Comment (dmfrey) [reply intermediate/testcontainers-rabbitmq/src/test/java/org/springframework/integration/samples/testcontainersrabbitmq/IntegrationConfigTest.java]: removed

File: intermediate/testcontainers-rabbitmq/src/test/java/org/springframework/integration/samples/testcontainersrabbitmq/IntegrationConfigTest.java [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+@SpringIntegrationTest
+@Import({ Receiver.class, IntegrationConfigTest.Config.class })
+class IntegrationConfigTest {
```
[the remark this answers is not in the document]
Comment (dmfrey) [reply intermediate/testcontainers-rabbitmq/src/test/java/org/springframework/integration/samples/testcontainersrabbitmq/IntegrationConfigTest.java]: ah, I do as well. It was generated by intellij this way. I've renamed it.

File: intermediate/testcontainers-rabbitmq/src/test/java/org/springframework/integration/samples/testcontainersrabbitmq/IntegrationConfigTest.java [hunk 10]
```diff
@@ -0,0 +1,93 @@
...
+        Binding resultsBinding( TopicExchange topicExchange, Queue resultsQueue ) {
+
+            return BindingBuilder.bind( resultsQueue )
+                    .to( topicExchange )
+                    .with( RESULTS_ROUTING_KEY );
+        }
+
+    }
+
+}
```
[the remark this answers is not in the document]
Comment (dmfrey) [reply intermediate/testcontainers-rabbitmq/src/test/java/org/springframework/integration/samples/testcontainersrabbitmq/IntegrationConfigTest.java]: added

File: intermediate/testcontainers-rabbitmq/src/test/java/org/springframework/integration/samples/testcontainersrabbitmq/TestcontainersRabbitmqApplicationTests.java (hunk 1 above)
Comment (dmfrey) [reply intermediate/testcontainers-rabbitmq/src/test/java/org/springframework/integration/samples/testcontainersrabbitmq/TestcontainersRabbitmqApplicationTests.java]: deleted

File: intermediate/testcontainers-rabbitmq/src/main/java/org/springframework/integration/samples/testcontainersrabbitmq/Request.java [hunk 11]
```diff
@@ -0,0 +1,14 @@
...
+import lombok.Value;
+
+import java.util.Optional;
+import java.util.UUID;
+
+@Value
+public class Request {
+
+    private final UUID id;
+    private final Optional<Integer> messageId;
```
[the remark this answers is not in the document]
Comment (dmfrey) [reply intermediate/testcontainers-rabbitmq/src/main/java/org/springframework/integration/samples/testcontainersrabbitmq/Request.java]: @artembilan do you mean to just separate each field with a blank line?

[the remark this answers is not in the document]
Comment (dmfrey) [reply intermediate/testcontainers-rabbitmq/src/main/java/org/springframework/integration/samples/testcontainersrabbitmq/Request.java]: i've put blank lines between them

File: intermediate/testcontainers-rabbitmq/src/main/java/org/springframework/integration/samples/testcontainersrabbitmq/RabbitConfig.java [hunk 12]
```diff
@@ -0,0 +1,44 @@
...
+public class RabbitConfig {
+
+    @Bean
+    public Jackson2JsonMessageConverter jsonMessageConverter( final ObjectMapper objectMapper ) {
+
+        return new Jackson2JsonMessageConverter( objectMapper );
+    }
+
+    @Bean
+    public RabbitTemplate amqpTemplate( ConnectionFactory connectionFactory, Jackson2JsonMessageConverter jsonMessageConverter ) {
```
[the remark this answers is not in the document]
Comment (dmfrey) [reply intermediate/testcontainers-rabbitmq/src/main/java/org/springframework/integration/samples/testcontainersrabbitmq/RabbitConfig.java]: Looks like I could eliminate this config altogether

File: intermediate/testcontainers-rabbitmq/src/main/java/org/springframework/integration/samples/testcontainersrabbitmq/RabbitConfig.java [hunk 13]
```diff
@@ -0,0 +1,44 @@
...
+        rabbitTemplate.setReplyAddress( "downstream.results" );
+        rabbitTemplate.setReplyTimeout( 10000 );
+        rabbitTemplate.setCorrelationKey( "request-correlation-id" );
+        rabbitTemplate.setUseDirectReplyToContainer( false );
+
+        return rabbitTemplate;
+    }
+
+    @Bean
+    public SimpleMessageListenerContainer replyListenerContainer( ConnectionFactory connectionFactory, RabbitTemplate amqpTemplate ) {
```
[the remark this answers is not in the document]
Comment (dmfrey) [reply intermediate/testcontainers-rabbitmq/src/main/java/org/springframework/integration/samples/testcontainersrabbitmq/RabbitConfig.java]: Looks like I could eliminate this config altogether

File: intermediate/testcontainers-rabbitmq/src/main/java/org/springframework/integration/samples/testcontainersrabbitmq/TestcontainersRabbitmqApplication.java [hunk 14]
```diff
@@ -0,0 +1,13 @@
+package org.springframework.integration.samples.testcontainersrabbitmq;
+
+import org.springframework.boot.SpringApplication;
+import org.springframework.boot.autoconfigure.SpringBootApplication;
+
+@SpringBootApplication
```
[the remark this answers is not in the document]
Comment (dmfrey) [reply intermediate/testcontainers-rabbitmq/src/main/java/org/springframework/integration/samples/testcontainersrabbitmq/TestcontainersRabbitmqApplication.java]: looks like it's not needed. test still passes by removing this rabbit config class

File: intermediate/testcontainers-rabbitmq/src/main/resources/application.yml [hunk 15]
```diff
@@ -0,0 +1 @@
+
```
[the remark this answers is not in the document]
Comment (dmfrey) [reply intermediate/testcontainers-rabbitmq/src/main/resources/application.yml]: no further configuration is needed

File: build.gradle [hunk 16]
```diff
@@ -917,6 +919,45 @@ project('tcp-client-server') {
...
+	dependencies {
+		compile "org.springframework.boot:spring-boot-starter-integration"
+		compile "org.springframework.integration:spring-integration-amqp"
+		compile "org.springframework.boot:spring-boot-starter-json"
+
+		testCompile('org.springframework.boot:spring-boot-starter-test') {
+			exclude group: 'org.junit.vintage', module: 'junit-vintage-engine'
+		}
+		testCompile 'org.springframework.integration:spring-integration-test'
+		testCompile 'org.springframework.integration:spring-integration-test-support'
```
Comment (artembilan) [inline build.gradle]: I don't think we need this one.
It has to come transitively from the `spring-integration-test`.

Comment (dmfrey) [reply build.gradle]: removed

**Comment by dmfrey** [comment]
@artembilan I've generated the pom with the gradle task and pushed up that commit

**Comment by artembilan** [comment]
Pulling it locally for testing on Windows 😄

**Comment by artembilan** [comment]
@dmfrey ,

Not a big deal, but we prefer tabs VS spaces for indents.

I'm fixing on merge...

**Comment by artembilan** [comment]
I also added a `Jackson2JsonMessageConverter` to avoid Java serialization and really get a gain from that `spring-boot-starter-json`.

Merging... Looks like Docker works well on my Windows! 😄

**Comment by artembilan** [comment]
Merged as https://github.com/spring-projects/spring-integration-samples/commit/427f28642ba9ac7ec612e88b3c4ae2420a959b0f after some clean up and improvements.

@dmfrey , 

thank you for contribution; looking forward for more!

**Comment by dmfrey** [comment]
> @dmfrey ,
> 
> Not a big deal, but we prefer tabs VS spaces for indents.
> 
> I'm fixing on merge...

Could this be added to a git preference?

**Comment by dmfrey** [comment]
> Merged as [427f286](https://github.com/spring-projects/spring-integration-samples/commit/427f28642ba9ac7ec612e88b3c4ae2420a959b0f) after some clean up and improvements.
> 
> @dmfrey ,
> 
> thank you for contribution; looking forward for more!

@artembilan no problem. I hope to be able to contribute more.

**Pull request closed without merging**